### PR TITLE
change drake geometry to header only inline methods.  

### PR DESCRIPTION
**Pull request by RussTedrake**
Looking quickly at:  https://travis-ci.org/RobotLocomotion/drake/builds , it looks like our build time perhaps already increased by ~20% with the last few template commits (starting with twip) -- but that also includes some new externals.

@psiorx -- might you be able to do a little exploration on the build servers to extract some numbers on the change in build time and object file size?

### CODE REVIEW

**Comment by RussTedrake** [comment]
looking only at travis, the build time seems comparable to before this pull.

**Comment by psiorx** [comment]
Just looked into this and found something interesting.  

It seems like the drakeJoints CRTP commit is the culprit.  It substantially increased both build time and object file size.  

I believe this is mostly due to the fact that drakeGradientUtil was converted to header only as part of that branch.  This forces all the code that uses drakeGradientUtil to both compile and include the inlined methods from it.

Here's some of the data:

```
MASTER
Build time: 5m6.137s
libdrakeJoints.so: 580K
libdrakeRBM.so: 1000K

8/25 2e4b393 <---drakeJoints-templated
Build time: 5m7.130s
libdrakeJoints.so: 580K	
libdrakeRBM.so: 1000K	

8/25 fbb3120
Build time: 4m12.204s
libdrakeJoints.so: 112K	
libdrakeRBM.so: 620K	

8/24 fda0e68
Build time: 4m12.045s

8/23 90b52ec
Build time: 4m14.033s

8/21 1f5035a
Build time: 4m15.257s

8/17 5b4cc0c <---- jc-twip
Build time: 4m15.257s

```

In summary, build time increased by roughly 20%, the binary size of libdrakeRBM increased by about 60%, and libdrakeJoints increased by 450% (!!) following this commit.

**Comment by psiorx** [comment]
Just for fun, I ran the same numbers against this branch:

```
8/26 02bc6b8 <----- RussTedrake:header_only_geom_util
Build time: 7m16.803s
libdrakeRBM.so: 1.1M	
```
That's quite a substantial jump (~40%) in build time.  Yikes!

**Comment by RussTedrake** [comment]
Thanks John.

It happens I just got finished reading about some of the other perils of inlined files.  We talked about:
- compile time

- code bloat (which can slow performance if you have more cache misses / paging at runtime). 
- longer compile time during development (editing headers means recompiling lots of targets in addition to relinking)
Other notes:
- a lot of debuggers will have trouble with inlines (e.g. can’t set breakpoints).  or they simply disable inlines for debug builds.
- the inline flag can be ignored by compilers (presumably not leading to different behavior, but certainly different object files).
- can’t provide binary-only library upgrades (e.g. to external consumers of drake)

I think drakeGeometryUtil might still be a good canddiate for headers — but probably moreso when we zap all of the gradient code in favor of the templated autodiff code.  But it looks like we need to be pretty careful before moving too many things over.

**Comment by RussTedrake** [comment]
give the stats, I think we should not merge this pull request, and should instead move methods to headers slowly and as necessary when we template them out.  For instance, in drakeGeometryUtil, we should move the actual geometry conversions over as we template them, but delete the user gradient versions instead of bringing them to the header file as well.  Will close this.

**Comment by tkoolen** [comment]
I kind of need this, so I'm going to argue for reopening.

> It seems like the drakeJoints CRTP commit is the culprit. It substantially increased both build time and object file size.
I believe this is mostly due to the fact that drakeGradientUtil was converted to header only as part of that branch. This forces all the code that uses drakeGradientUtil to both compile and include the inlined methods from it.

Well, the other major change introduced by the CRTP thing was doubling the number of DrakeJoint methods, i.e. adding the variants that accept AutoDiffScalar. These methods also require additional instantiations of drakeGeometryUtil methods. So I don't think that assessment is completely fair and should be reevaluated by not compiling the AutoDiffScalar variants and measuring the compile time.

> It happens I just got finished reading about some of the other perils of inlined files.

Now we're conflating two different things again: use of the inline keyword and putting definitions of templated functions in the header. Could we do a proper experiment where the only difference is whether or not templated methods are implemented in the header file?

You guys keep talking about binary size by the way, and I'm wondering why. I really don't care if a library is 500 kB or 1 MB myself. Also, you're not counting the reduction in size of libdrakeGeometryUtil to zero.

**Comment by tkoolen** [comment]
The non-templated methods should remain in the cpp file by the way.
